### PR TITLE
use promise.

### PR DESCRIPTION
**Pull request by tokuhirom**

### CODE REVIEW

File: lib/Crust/Middleware/AccessLog.pm6 [hunk 1]
```diff
@@ -38,21 +38,23 @@ my sub content-length(@res) {
...
-        $logger = sub ($s) { %env<p6sgi.errors>.print($s) };
-    }
+    start {
+        my $t0 = DateTime.now.Instant;
+        my @res = $.app()(%env).result;
+
+        # '%h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-agent}i"'
+        my $logger = $.logger;
+        if !$logger.defined {
+            $logger = sub ($s) { %env<p6sgi.errors>.print($s) };
```
Comment (astj) [inline lib/Crust/Middleware/AccessLog.pm6]: apps (/middlewares) should emit errors by `%env<p6sgi.errors>.emit($s)`. ref: #94

File: lib/Crust/Middleware/StackTrace.pm6 [hunk 2]
```diff
@@ -7,27 +7,30 @@ unit class Crust::Middleware::StackTrace is Crust::Middleware;
...
+
+            CATCH {
+                my $trace = .backtrace;
+                default {
+                    my $text = @$trace.map({ .Str.trim }).join("\n");
+                    my $html = $trace.as-html;
+                    %env<crust.stacktrace.text> = $text;
+                    %env<crust.stacktrace.html> = $html;
+
+                    %env<p6sgi.errors>.print($text) unless $.no-print-errors;
```
Comment (astj) [inline lib/Crust/Middleware/StackTrace.pm6]: #94

File: lib/Crust/Middleware/ContentLength.pm6 [hunk 3]
```diff
@@ -5,20 +5,22 @@ use Crust::Utils;
...
-        !%headers<Transfer-Encoding>.defined &&
-        (my $content-length = content-length(@ret[2])).defined
-    ) {
-        %headers<Content-Length> = $content-length;
-    }
-    @ret[1] = [%headers];
-
-    return @ret;
+    return start {
+        my @ret = await $.app()(%env);
```
Comment (astj) [inline lib/Crust/Middleware/ContentLength.pm6]: [nit]
We can use `Promise.then` here?

File: lib/Crust/Middleware/Static.pm6 [hunk 4]
```diff
@@ -51,6 +51,6 @@ method !handle-static(%env) {
     }
 
     temp %env<PATH_INFO> = $path;
-    my @res = $!file.(%env);
+    my @res = await $!file.(%env);
```
Comment (astj) [inline lib/Crust/Middleware/Static.pm6]: `$!file.(%env)` should be a Promise, so just returning `$!file.(%env)` seems enough.

Comment (tokuhirom) [inline lib/Crust/Middleware/Static.pm6]: * M::Static needs to check the status code In the CALL-ME method.
* I want to rewrite PATH_INFO in the scope.

File: lib/Crust/Middleware/ContentLength.pm6 (hunk 3 above)
Comment (tokuhirom) [inline lib/Crust/Middleware/ContentLength.pm6]: hmm. we should discuss after merging :P

File: lib/Crust/Middleware/Static.pm6 (hunk 4 above)
Comment (astj) [inline lib/Crust/Middleware/Static.pm6]: Oh, I understand. 🙆‍♂️

**Pull request merged**